### PR TITLE
feat: replace any env vars in npmrc

### PR DESCRIPTION
**Pull request by ctaepper**
until now, only `NPM_TOKEN` is used from environment variables (see #851)
this PR uses basically the same approach as [yarn](https://github.com/yarnpkg/yarn/pull/1207/files)

### CODE REVIEW

**Comment by rarkins** [comment]
@ctaepper thanks for your contribution. Could you elaborate a little on the use case, so I fully understand it? I assume this must be for your own self-hosted bot, for example. Is it intended to be for `.npmrc` files in the repositories you are renovating?

In its current form I think this opens up a security hole. e.g. it would allow someone to use a `.npmrc` in a repository I don't trust to them extra variables from my process.env. If it's only intended to be used where you trust ever committer to every repository you run, then we can maybe consider some other way to allow the administrator to decide to enable/disable. Please tell me more about your use case first so I can suggest the best approach.

**Comment by ctaepper** [comment]
we have a couple of projects using one or more private npm scopes, or even private npm registries... only one `NPM_TOKEN` wont do it.

example npmrc file:
```
@custom:registry=https://custom-registry.com/
//custom-registry.com/:_password=${ARTIFACTORY_PASSWORD}
//custom-registry.com/:username=${ARTIFACTORY_USER}
//custom-registry.com/:email=${ARTIFACTORY_EMAIL}
//custom-registry.com/:always-auth=true
```
so yes, this is especially useful for self hosted version. maybe this feature could only be enabled for selfhosted versions, by environment variable for example. or even better, it could be enabled by default, and you disable this only in your public service

**Comment by rarkins** [comment]
So your use case is:
- Self hosted Renovate bot
- Multiple repositories
- Repositories already contain `.npmrc` files with env variable injection
- You will configure your Renovate bot host with the right environment variables and they will be injected into the `.npmrc` by the bot
?

I have follow-up questions, but want to make sure I understand this part first. Due to the security risk here (this is only acceptable if you trust all committers to all repositories) I would definitely want to disable this by default and make it opt-in. Even in self-hosted, it doesn't mean everyone trusts everyone.

**Comment by ctaepper** [comment]
exactly what you said, this is my case

**Comment by rarkins** [comment]
@ctaepper do you have lock files to consider too? e.g. `yarn.lock` or `package-lock.json`? I'm guessing it's yarn. So in that case it should be "in sync" with Renovate if the env is also exposed to the shell where `yarn` runs. I need to double check but I'm pretty sure though that I already "clean" the env before calling yarn, so that will need to be part of this PR too or it won't fully work. I guess the same variable could control both, something like `exposeEnv` true/false.

**Comment by ctaepper** [comment]
the very current use case is actually 3 repositories using monorepo approach with lerna and npm with package-locks.

**Comment by rarkins** [comment]
> npm with package-locks

So npm supports this command substitution too, while then every developer and test machine is configured to have these artifactory env vars set?

**Comment by ctaepper** [comment]
yes, developers working on those projects have to expose those env vars on their machines

**Comment by rarkins** [comment]
@ctaepper thanks, now I think I understand the use case. Can you fix the test coverage as-is? Afterwards, I'll add the new config option I want to control its enablement, plus the logic to expose full env to child_process when npm and yarn are called.

**Comment by ctaepper** [comment]
@rarkins while digging through the code, I think I found another issue, please correct me if I'm wrong:
I hit the problem that you mentioned earlier about the lockfiles, so i exposed the full env to the `npm install` child processes and expected everything to work. it didn't :)

we are making use of lerna monorepo manager and we have a coupe of dependencies between the lerna packages inside the repo, e.g. a config package which a direct dependency of some of the other packages. those internal dependencies are marked as
```
"dependencies": {
  "@myscope/config": "*"
}
```
before publishing, lerna replaces `"*"` with the actual latest version of that dependency.

our current case is, that the project is under heavy development, however, we haven't hat a actual release yet, so nothing is published to npm until we are ready to go live.

from what i understand about renovate
- the process detects if there is a lockfile
- if so, it takes the according package.json and runs `npm install --package-lock-only` in order to generate a lockfile

however, if this package.json contains `"@myscope/config": "*"`, this install will fail, because npm is not aware of this package.

the *lerna way* to do that is to use `lerna bootstrap` within the monorepo root. lerna will then, for each package,
- create a temporary new package.json
- remove all internal dependencies
- run npm install
- restore the original package.json

this way, the npm install will never hit the internal dependencies

**beware:** this is only applicable to `npm`. with `yarn (workspaces)` lerna will leave those action to yarn itsself

----
bottom line is that even if I workaround (or PR) the environment variable problems, I can't use renovate because it doesn't handle internal monorepo dependencies correctly. or am I missing something?

**Comment by rarkins** [comment]
@ctaepper thanks for looking into it in advance, and also for the detailed explanation!

The only reason we spawn `npm` is if we need to update the `package-lock.json`. If there is another way, then we can consider it.

In this case, is there one lock file for the whole lerna (i.e. like Yarn Workspaces), or is it one per-`package.json`?

If I understand you correctly, then we need similar approach as for Yarn Workspaces, where we have the logic: if any `package.json` is updated then then write out all the `package.json` files in the workspaces and then run `yarn install` once from the workspaces root.

In this case, it sounds like the above except we'd need to run `lerna bootstrap` from the root? 

Assuming you need to run it from a globally-installed `lerna`, do you have any way to ensure the whole team is running the exact same version of that?

As you can probably guess, when supporting thousands of repositories in the App I don't want to actually install *any* `node_modules`, but if this is for your own hosted instance then there's no reason to block the idea for load concern reasons. If you end up running this a lot then we probably want to *skip* cleaning up the node_modules between bootstraps and even between runs. Today Renovate uses temporary directories with automatic cleanup deliberately to avoid any problems with stale caches, etc - but it's a conservative approach and in most cases keeping the existing `node_modules` in place is safe. In earlier versions of `npm`.. not so safe, in case the modules were corrupted.

**Comment by ctaepper** [comment]
@rarkins with yarn it is always a single lockfile

with npm, lerna has a `--hoist` option. if activated there will be only on lockfile in the repo root. if not activated (our current case) we have one lockfile for each monorepo package

lerna itsself is added to the devdependencies of the repo root (by default on `lerna init`). so any lerna command acutally ~~resides~~ should reside within a npm script in the root package json in order to make sure every developer uses the same version. theoretically this is not a must-have. one could globally install lerna and use the commands directly, however, as with many other tooling dependencies, this is never good practice...

so yes, the solution would be to run lerna bootstrap in the repo root in order to generate proper lockfiles (whereever they are) while completely ignoring all inner-monorepo-dependencies

I tried out `lerna bootstrap -- --package-lock-only`  and it works

**Comment by ctaepper** [comment]
also, I tried to add some tests, however, this is my first contact with jest, so I don't have the hang of it just yet... maybe you can point me to those 100% ?

**Comment by rarkins** [comment]
> the solution would be to run lerna bootstrap in the repo root

I guess we could ourselves add `lerna` as a dependency of `renovate` - like we do with `npm`, `yarn` and `pnpm` - and then run it from `node_modules` to save having to `lerna init`? The only risk is if we get ahead of your project and somehow the newer lerna is not compatible with your old lerna.. however in such a case you could use an older Renovate until you fix the problem and update your project's `lerna` to match.

Please tell me if I've understood this correctly:

- `lerna` would be added as a production dependency of `renovate`, so that it's versioned and always available
- Instead of running `npm install` locally on each `package.json` that's changed, we should instead run `lerna bootstrap -- --package-lock-only` in the root of the repo once per branch/PR (similar to how we do with `yarn install` for Workspaces)
- Instead of only checking/updating the one `yarn.lock`, we need to check for any/all `package-lock.json` files being changed and add them all to the commit if changed

**Comment by ctaepper** [comment]
> to save having to `lerna init`

`lerna init` should never be called by renovate

---
`given`
- renovate detects `lerna.json`
- and renovates any dependency in root or sub packages

`when`
- lerna detects `package-lock.json` in project root (this will always be there if lockfiles are enabled)

`then`
- renovate should assume lerna is configured to use `npm`
- renovate should have a lerna executable available
  - with the lerna version specified in `devDependencies` of the root `package.json`
  - if not found in `devDependencies` (very bad practice) it should fallback to a version specified by renovate itsself (maybe add a warning to the onboarding PR)
- renovate should execute `npm install --package-lock-only` in order to handle the root dependencies
- renovate should execute `lerna bootstrap -- --package-lock-only` in order to handle the subpackage dependencies


depending on the `hoist` option in lerna.json this will result
- `hoist: false (default)` one `package-lock.json` in project root for root devDependencies and one package-lock.json for each monorepo package
- `hoist: true` one updated `package-lock.json` in project root
  - unless there are version range conflicts within the monorepo packages
  - the version which is more "common" will be hoisted in the root `node_modules`, and the package with the conflicting version range will get a `package-lock.json` only containing the conflicting dependencies
---

bottom line: it doesnt matter which hoist option is used, there can be updated package-lock.json's in root and/or subpackages which should all be staged for commit

PS:
> renovate should assume lerna is configured to use `npm`

i'm sure you already have logic in place to differenciate between npm and yarn. on top of that, you could check if the property `npmClient` is present in `lerna.json`

**Comment by rarkins** [comment]
@ctaepper re: tests/coverage, you can run them locally with `yarn jest` or `yarn jest -u` if you need to update snapshots.Coverage for this file can be viewed in `coverage/lcov-report/lib/manager/npm/registry.js.html` or via codecov after a successful test on Travis. Right now the URL is here: https://codecov.io/gh/renovateapp/renovate/compare/1aad8ec053a343b3b16beb8c20fed68b9f150f3d...401059771afef9a8443218e6053453bef2246435/diff

you can choose to add more tests cases if those if statements are applicable, remove them if they're not applicable to us, or `// istanbul ignore if` them if they are applicable but too difficult to test.

**Comment by ctaepper** [comment]
@rarkins all cleaned up
if you want help regarding the lerna issues, I am willing to invest some time... can you point me into the right directions? like where could I start adding this logic?

**Comment by rarkins** [comment]
@ctaepper I'll add some extra code now to wrap this new behaviour in a config option - as well as using same option for exposing `env` when calling child processes. After that, this one can be merged - although doesn't help you too much on its own yet. We can then continue the `lerna bootstrap` part in #1441
